java/ql/src/experimental/Security/CWE/CWE-555/CredentialsInPropertiesFile.ql

```diff
@@ -0,0 +1,87 @@
+/**
+ * @name Cleartext Credentials in Properties File
+ * @description Finds cleartext credentials in Java properties files.
+ * @kind problem
+ * @id java/credentials-in-properties
+ * @tags security
+ *       external/cwe/cwe-555
+ *       external/cwe/cwe-256
+ *       external/cwe/cwe-260
+ */
+
+import java
+import semmle.code.configfiles.ConfigFiles
+
+private string suspicious() {
+  result = "%password%" or
+  result = "%passwd%" or
+  result = "%account%" or
+  result = "%accnt%" or
+  result = "%credential%" or
+  result = "%token%" or
+  result = "%secret%" or
+  result = "%access%key%"
+}
+
+private string nonSuspicious() {
+  result = "%hashed%" or
+  result = "%encrypted%" or
+  result = "%crypt%"
+}
+
+/** Holds if the value is not cleartext credentials. */
+bindingset[value]
+predicate isNotCleartextCredentials(string value) {
+  value = "" // Empty string
+  or
+  value.length() < 7 // Typical credentials are no less than 6 characters
+  or
+  value.matches("% %") // Sentences containing spaces
+  or
+  value.regexpMatch(".*[^a-zA-Z\\d]{3,}.*") // Contain repeated non-alphanumeric characters such as a fake password pass**** or ????
+  or
+  value.matches("@%") // Starts with the "@" sign
+  or
+  value.regexpMatch("\\$\\{.*\\}") // Variable placeholder ${credentials}
+  or
+  value.matches("%=") // A basic check of encrypted credentials ending with padding characters
+  or
+  value.matches("ENC(%)") // Encrypted value
+  or
+  value.toLowerCase().matches(suspicious()) // Could be message properties or fake passwords
+}
+
+/**
+ * Holds if the credentials are in a non-production properties file indicated by:
+ *    a) in a non-production directory
+ *    b) with a non-production file name
+ */
+predicate isNonProdCredentials(CredentialsConfig cc) {
+  cc.getFile().getAbsolutePath().matches(["%dev%", "%test%", "%sample%"]) and
+  not cc.getFile().getAbsolutePath().matches("%codeql%") // CodeQL test cases
```

Comment:
Remove this, it's ok if we identify our own tests

Reply:
Thanks @smowton for reviewing this PR.

The following code snippet is to exclude non-production properties files to reduce FPs.
```
cc.getFile().getAbsolutePath().matches(["%dev%", "%test%", "%sample%"])
```

And the following code snippet is to include our own test cases under the <code>test/experimental/query-tests/security/CWE-555</code> directory, or else no result will be returned in <code>CredentialsInPropertiesFile.expected</code>. 

Do you want to remove both or just the <code>codeql</code> check? Please advise. Thanks.

Reply:
just remove `not cc.getFile().getAbsolutePath().matches("%codeql%")`. If there's a technical need to exclude something in one of our own test cases, use a `.ql` file in the test folder to achieve that.

Reply:
Thanks @smowton for the clarification. I've copied the <code>.ql</code> file to the test folder and removed the <code>isNonProdCredentials</code> check so that it can work with our test cases.